src/osd/PrimaryLogPG.cc

```diff
@@ -9976,6 +9978,264 @@ struct C_Flush : public Context {
+
+
+int PrimaryLogPG::start_dedup(OpRequestRef op, ObjectContextRef obc)
+{
+  bufferlist bl;
+  uint64_t cur_off = 0;
+  const object_info_t& oi = obc->obs.oi;
+  const hobject_t& soid = oi.soid;
+
+  if (!obc->is_blocked()) {
+    return -EINVAL;
+  }
+
+  /*
+   * The operations to make dedup chunks are tracked by a ManifestOp.
+   * This op will be finished if all the operations are completed.
+   */
+  ManifestOpRef mop(std::make_shared<ManifestOp>(nullptr, 0));
+
+  while (cur_off < oi.size) {
+    // cdc
+    vector<pair<uint64_t, uint64_t>> chunks;
+    int r = do_cdc(oi, bl, chunks);
+    if (r < 0) {
+      return r;
+    }
+    if (!chunks.size()) {
+      break;
+    }
+
+    // get fingerprint 
+    for (auto p : chunks) {
+      bufferlist chunk;
+      chunk.substr_of(bl, p.first, p.second);
+      hobject_t target = get_fpoid_from_chunk(soid, chunk);
+      if (target == hobject_t()) {
+	dout(0) << " fingerprint oid is null " << dendl;
+	break;
+      }
+
+      // make a create_or_get_ref op
+      bufferlist t;
+      ObjectOperation obj_op;
+      cls_cas_chunk_create_or_get_ref_op get_call;
+      get_call.source = soid.get_head();
+      get_call.data = chunk;
+      ::encode(get_call, t);
+      obj_op.call("cas", "chunk_create_or_get_ref", t);
+
+      // issue create_or_get_ref_op
+      C_SetDedupChunks *fin = new C_SetDedupChunks(this, soid, get_last_peering_reset());
+      fin->offset = p.first;
+      object_locator_t oloc(target);
+      unsigned flags = CEPH_OSD_FLAG_IGNORE_CACHE | CEPH_OSD_FLAG_IGNORE_OVERLAY |
+			CEPH_OSD_FLAG_RWORDERED;
+      ceph_tid_t tid = osd->objecter->mutate(
+	  target.oid, oloc, obj_op, SnapContext(),
+	  ceph::real_clock::from_ceph_timespec(obc->obs.oi.mtime),
+	  flags, new C_OnFinisher(fin, osd->get_objecter_finisher(get_pg_shard())));
+      fin->tid = tid;
+      mop->tids[p.first] = tid;
+      mop->chunks[target] = make_pair(p.first, p.second);
+      mop->num_chunks++;
+      dout(10) << __func__ << " oid: " << soid << " tid: " << tid 
+	      << " target: " << target << " offset: " << p.first
+	      << " length: " << p.second << dendl;
+    }
+    
+    cur_off += bl.length();
+  }
+
+  if (mop->tids.size()) {
+    manifest_ops[soid] = mop;
+    manifest_ops[soid]->op = op;
+  } 
+
+  return -EINPROGRESS;
+}
+
+int PrimaryLogPG::do_cdc(const object_info_t& oi, bufferlist& bl, vector<pair<uint64_t, uint64_t>>& chunks)
+{
+  uint64_t cur_off = 0;
+  string chunk_algo = pool.info.get_dedup_chunk_algorithm_name();
+  int64_t chunk_size = pool.info.get_dedup_cdc_chunk_size();
+  uint64_t max_window_size = static_cast<uint64_t>(pool.info.get_dedup_cdc_window_size());
+
+  std::unique_ptr<CDC> cdc = CDC::create(chunk_algo, cbits(chunk_size)-1);
+  if (!cdc) {
+    dout(0) << __func__ << " unrecognized chunk-algorithm " << dendl;
+    return -EINVAL;
+  }
+  while (cur_off < oi.size && cur_off < max_window_size) {
+    bufferlist chunk_data;
+    // TODO: Do we need to support EC?
```

Comment:
I think it's ok for now not to.  Just make sure to return an error code and add a test validating the error code behavior.

Reply:
Probably the mon shouldn't allow an ec pool to be set as the top dedup pool?

Reply:
That is, I think the pool settings added here should be invalid on an EC pool.

Reply:
You left this as a TODO with a question.  Please change the comment to explain your current choice and reasons so that readers later won't have to guess.